From c6aaf4582bdbb9b4a9f6b072795a732afad2f22b Mon Sep 17 00:00:00 2001
From: Han Xu <han.xu@nxp.com>
Date: Fri, 21 Jul 2017 16:17:59 -0500
Subject: [PATCH 2184/5242] MLK-16060: mtd: fsl-quadspi: fix the unalignment
 issue for qspi

commit  e7bdef5e1a3575237cdcfe55f0b3f53630522107 from
https://source.codeaurora.org/external/imx/linux-imx.git

ARM64 platforms may access QSPI from non-64-bit-aligned address which
causes unalignment fault. Fixed the issue for AHB reading.

Signed-off-by: Han Xu <han.xu@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/mtd/spi-nor/fsl-quadspi.c |   21 ++++++++++++++++++---
 1 file changed, 18 insertions(+), 3 deletions(-)

diff --git a/drivers/mtd/spi-nor/fsl-quadspi.c b/drivers/mtd/spi-nor/fsl-quadspi.c
index dcff0da..b733a24 100644
--- a/drivers/mtd/spi-nor/fsl-quadspi.c
+++ b/drivers/mtd/spi-nor/fsl-quadspi.c
@@ -2,6 +2,7 @@
  * Freescale QuadSPI driver.
  *
  * Copyright (C) 2013-2016 Freescale Semiconductor, Inc.
+ * Copyright 2017 NXP
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
@@ -1005,6 +1006,8 @@ static ssize_t fsl_qspi_read(struct spi_nor *nor, loff_t from,
 {
 	struct fsl_qspi *q = nor->priv;
 	u8 cmd = nor->read_opcode;
+	u8 tmp[8] = {0};
+	int i, j;
 
 	/* if necessary,ioremap buffer before AHB read, */
 	if (!q->ahb_addr) {
@@ -1039,9 +1042,21 @@ static ssize_t fsl_qspi_read(struct spi_nor *nor, loff_t from,
 		cmd, q->ahb_addr + q->chip_base_addr + from - q->memmap_offs,
 		len);
 
-	/* Read out the data directly from the AHB buffer.*/
-	memcpy(buf, q->ahb_addr + q->chip_base_addr + from - q->memmap_offs,
-		len);
+	/* For non-8-byte alignment cases */
+	if (from % 8) {
+		j = 8 - (from & 0x7);
+		for (i = 0; i < j; ++i) {
+			memcpy(tmp + i, q->ahb_addr + q->chip_base_addr + from
+			       - q->memmap_offs + i, 1);
+		}
+		memcpy(buf, tmp, j);
+		memcpy(buf + j, q->ahb_addr + q->chip_base_addr + from
+		       - q->memmap_offs + j, len - j);
+	} else {
+		/* Read out the data directly from the AHB buffer.*/
+		memcpy(buf, q->ahb_addr + q->chip_base_addr + from
+		       - q->memmap_offs, len);
+	}
 
 	return len;
 }
-- 
1.7.9.5

